torch/distributed/_checkpoint/state_dict_loader.py

```diff
@@ -0,0 +1,250 @@
+
+import io
+from typing import Any, Dict, List, Optional, Tuple
+
+import torch
+from torch.distributed._shard.sharded_tensor import (
+    ShardedTensor,
+    ShardedTensorMetadata
+)
+from torch.distributed._shard.sharding_spec._internals import (
+    validate_non_overlapping_shards_metadata,
+    _check_shard_metadata_pair_overlap,
+)
+
+from .metadata import (
+    BytesReadRequest,
+    TensorReadRequest,
+    Metadata,
+    ShardedTensorStorageMetadata,
+    TensorStorageMetadata,
+)
+from .resharding import (
+    _prepare_sharded_tensor_read,
+    _shards_get_overlap_region_wrt_saved_tensor
+)
+from .storage_reader import StorageReader
+
+
+def _reshard_and_prepare_read_request(
+    state_dict: Dict[str, Any], metadata_from_storage: Metadata
+) -> Tuple[List[BytesReadRequest], List[TensorReadRequest]]:
+    """
+    Use the loaded metadata and the current state dict to map the saved tensors to current tensor
+    """
+    tensor_read_requests = []
+    bytes_read_requests = []
+    for fqn, obj in state_dict.items():
+        if isinstance(obj, torch.Tensor):
+            tensor = obj.detach()
+            md = metadata_from_storage.state_dict_metadata[fqn]
+            if isinstance(md, TensorStorageMetadata):
+                rr = TensorReadRequest(
+                    tensor=tensor,
+                    storage_key=fqn,
+                    offsets=tuple([0] * len(tensor.size())),
+                    lengths=md.size,
+                )
+
+                tensor_read_requests.append(rr)
+            else:
+                raise ValueError(
+                    f"Invalid checkpoint metadata for {fqn}, " +
+                    "expected TensorStorageMetadata but found {type(md)}"
+                )
+        elif isinstance(obj, ShardedTensor):
```

Comment:
same here, let's do `isinstance` check for ShardedTensor first then torch.Tensor second to avoid issues when torch.Tensor subclass PR landed